lib/rust/parser/jni/src/lib.rs

```diff
@@ -16,6 +16,7 @@ use jni::objects::JClass;
 use jni::sys::jobject;
 use jni::sys::jstring;
 use jni::JNIEnv;
+use std::cell::UnsafeCell;
```

Comment:
"Unsafe" Rust!? Wouldn't it be better to use some "safe" data structure? `Vector`?

Reply:
Globally-scoped data (such as thread-local state in this case) is not inherently `safe` to mutate in Rust; some kind of `Cell` is needed to allow it. The options here are `RefCell` (which is essentially a runtime-checked *single-threaded* mutex) or `UnsafeCell` (which allows access without any runtime checks).
The soundness requirement of `UnsafeCell` is that it is not used to establish multiple references to its contained data concurrently with any of the references being mutators; in this case that property is ensured by limiting the scope of borrows from `STATE`--no borrow derived from `STATE` escapes the function that obtains it. Rust lifetimes make this easy to enforce, so we don't need the runtime overhead of the assertion that `RefCell` would use to make this `safe`.